Enable autovacuum, to deal with 'template0'

hlinnaka opened the pull request:
There are two commits here. The first one, to track "oldest xmin" in segments, is perhaps the  more interesting one. See commit messages, for details.

This currently fails the `vacuum_full_recently_dead_tuple_due_to_distributed_snapshot`, but that's OK: The test was written to test that VACUUM in utility mode is conservative, and won't remove tuples that are in fact dead. But with this PR, it can do so safely. So the test tested the overly-conservative approach that it used to do. We should probably change it or write a new one, to test that VACUUM in utility mode doesn't incorrectly remove tuples that are still visible to a distributed snapshot.

hlinnaka commented:
Oh, and the third commit adds a GUC that I found really useful while testing this. To rack up the XID counter in a segment, I used `PGOPTIONS='-c gp_session_role=utility' PGPORT=40000 pgbench -f burn-xids.sql -n postgres`, with this `burn-xids.sql` script:
```
set debug_burn_xids=on;
select txid_current();
```

ashwinstar commented on src/backend/access/transam/distributedlog.c:
```diff
@@ -73,6 +77,138 @@ static bool DistributedLog_PagePrecedes(int page1, int page2);
+
+		if (page != currPage)
+		{
+			slotno = SimpleLruReadPage(DistributedLogCtl, page, true, xid);
```
We should leverage the process local cache here `LocalDistribXactCache_CommittedFind()` to help find localXid to distributedXid mapping faster if possible.

hlinnaka replied:
Hmm, I don't think that's useful here. I suppose we could use it, instead of reading the SLRU page, if the value happens to be in the cache. But: if there are any XIDs that are not found in the cache, we'd have to hit the SLRU anyway. And because we're scanning the SLRU sequentially, it's faster than using the cache, if we can advance the value more than 1-2 xids. And we'll need to grab the DistributedLogControlLock anyway, in order to update the oldestxmin value kept in shared memory.

ashwinstar commented on src/backend/access/transam/distributedlog.c:
```diff
@@ -73,6 +77,138 @@ static bool DistributedLog_PagePrecedes(int page1, int page2);
+	{
+		/*
+		 * If we haven't "seen" any distributed snapshots from the QD yet,
+		 * initialize oldest xmin to the oldest datfrozenxid in this segment.
```
ya starting from `datfrozenxid` seems too much processing, must preserve the same across restarts.

ashwinstar commented on src/backend/access/transam/distributedlog.c:
```diff
@@ -73,6 +77,138 @@ static bool DistributedLog_PagePrecedes(int page1, int page2);
+		 * (GetNewTransactionId() can call DistributedLog_Truncate(), while
+		 * already holding XidGenLock)
+		 */
+		LWLockRelease(DistributedLogControlLock);
```
Not fan of `LWLockRelease()` and `LWLockAcquire()` again, does it pose race here where other backend updates the `DistributedLogShared->oldestXmin` meanwhile. Seems mostly would just do duplicate work ?

hlinnaka replied:
Yeah, it's just duplicated work. I'll add a check, after re-acquiring the lock, to see if `DistributedLogShared->oldestXmin` was changed, while we were not holding the lock. That'll avoid scanning the SLRU, if another backend already did it.

ashwinstar commented on src/backend/access/transam/distributedlog.c:
```diff
@@ -73,6 +77,138 @@ static bool DistributedLog_PagePrecedes(int page1, int page2);
+		 * past it. Otherwise stop here.
+		 */
+		if (ptr->distribTimeStamp != distribTransactionTimeStamp ||
+			TransactionIdPrecedes(ptr->distribXid, xminAllDistributedSnapshots))
```
The approach seems lot more conservative but maybe fine in practice, as higher localXid can map to lower distribXid which can reclaimed but it stops here.

hlinnaka replied:
Yeah. That is the general tradeoff we make with OldestXmin. It's also possible with purely local XIDs, that OldestXmin is held back by one very old transaction, and we could safely remove/freeze some XIDs newer than OldestXmin, if we just kept track of snapshots in a more fine-grained fashion. HeapTupleSatisfiesVacuum() is called for every tuple during VACUUM, so it's important that the test against OldestXmin is simple and cheap.

ashwinstar commented on src/backend/storage/ipc/procarray.c:
```diff
@@ -1857,6 +1891,20 @@ GetSnapshotData(Snapshot snapshot)
+	 */
+	if (Gp_segment != -1)
+	{
+		DistributedSnapshot *ds = &snapshot->distribSnapshotWithLocalMapping.ds;
```
need validation if DistributedSnapshot is not present ?

ashwinstar commented on src/backend/access/transam/distributedlog.c:
```diff
@@ -73,6 +77,138 @@ static bool DistributedLog_PagePrecedes(int page1, int page2);
+	DistributedLogEntry *ptr;
+	DistributedLogEntry *entries;
+
+	LWLockAcquire(DistributedLogControlLock, LW_EXCLUSIVE);
```
Seems will introduce contention on `DistributedLogControlLock`.

xinzweb commented on src/backend/access/transam/varsup.c:
```diff
@@ -186,6 +186,21 @@ GetNewTransactionId(bool isSubXact)
 	 */
 	TransactionIdAdvance(ShmemVariableCache->nextXid);
 
+	if (Debug_burn_xids)
```
Shall we do the `Debug_burn_xids` before the `TransactionIdAdvance` so that it can wrap around?

xinzweb commented on src/backend/access/transam/varsup.c:
```diff
@@ -186,6 +186,21 @@ GetNewTransactionId(bool isSubXact)
+
+		xx = ShmemVariableCache->nextXid;
+
+		r = xx % 1024;
```
Not sure the logic using 1024, why not using `CLOG_XACTS_PER_PAGE`? Then, if we turn the `Debug_burn_xids`, each transaction will jump to end of the current clog page?

xinzweb commented on src/backend/access/transam/varsup.c:
```diff
@@ -186,6 +186,21 @@ GetNewTransactionId(bool isSubXact)
+	if (Debug_burn_xids)
+	{
+		TransactionId xx;
+		int			r;
```
Why not just use unit32 for the r? Under what condition, do we want this r to be negative? Thanks.

xinzweb commented:
@hlinnaka add test to verify autovacuum both positive and negative test cases like:
- autovacuum only kicked in for the `template0`
- autovacuum not triggered on `tempate1` and any other user databases like `postgres`

xinzweb commented on src/backend/postmaster/autovacuum.c:
```diff
@@ -1803,6 +1802,16 @@ get_database_list(void)
 		avw_dbase   *avdb;
 
+		/*
+		 * In GPDB, autovacuum is currently disabled, except for the
```
One possible solution would skip the shared relations when building the `table_oids` inside `do_autovacuum()`? In this case, we will not touch any shared relations and make this autovacuum fully localize to that give database (template0).

hlinnaka commented:
Pushed a new version, with some changes to the way the "OldestXmin" value in shared memory is initialized. I think this should address Ashwin's performance concerns on the first call.

On the first call after postmaster startup, like before, we now start scanning from the oldest datfrozenxid. But instead of scanning one XID at a time, we first advance one SLRU page at a time with 'SimpleLruPageExists`, skipping over pg_distributedlog pages that have already been truncated away. And to make sure that helps, we truncate pg_distributedlog more aggressively, whenever we find that oldestXmin has advanced to a new SLRU segment. That should greatly speed up the initialization after postmaster startup.

This doesn't change the fact that we now need to acquire DistributedLogControlLock every time a snapshot is acquired. That could become a problem, hurting concurrency. To alleviate that, we could make it work so that only one backend would try to advance the oldest xmin value at a time, and if the lock is taken because someone else is already advancing it, we would just use the old value.  But I'd like to keep this simple, until someone demonstrates that there is an actual problem. These things can easily be masked by much bigger overheads elsewhere.

ashwinstar commented:
On Tue, Mar 6, 2018 at 6:52 AM, Heikki Linnakangas <notifications@github.com
> wrote:

> Pushed a new version, with some changes to the way the "OldestXmin" value
> in shared memory is initialized. I think this should address Ashwin's
> performance concerns on the first call.
>
> On the first call after postmaster startup, like before, we now start
> scanning from the oldest datfrozenxid. But instead of scanning one XID at a
> time, we first advance one SLRU page at a time with 'SimpleLruPageExists`,
> skipping over pg_distributedlog pages that have already been truncated
> away. And to make sure that helps, we truncate pg_distributedlog more
> aggressively, whenever we find that oldestXmin has advanced to a new SLRU
> segment. That should greatly speed up the initialization after postmaster
> startup.
>
Lovely !!! Really liked the simplicity of avoiding the performance penalty
by aggressive distributed log truncation.
Actually, it may turn out to be performance boost for regular queries as
well, as might result in faster dxid->lxid lookups.

> This doesn't change the fact that we now need to acquire
> DistributedLogControlLock every time a snapshot is acquired. That could
> become a problem, hurting concurrency. To alleviate that, we could make it
> work so that only one backend would try to advance the oldest xmin value at
> a time, and if the lock is taken because someone else is already advancing
> it, we would just use the old value. But I'd like to keep this simple,
> until someone demonstrates that there is an actual problem. These things
> can easily be masked by much bigger overheads elsewhere.
>
Tricky part is since there exist no performance pipeline to measure and
validate the performance regressions, someone else will demonstrate the
problem seems unlikely :-) Will trying just some simple pg_bench
concurrency runs will give some rough idea, if this affects or not ?

ashwinstar commented:
Just thinking is there possibility to land in weird conditions for incorrectly missing distributed log files, as now its not treated as ERROR condition but rather driving decisions ?

asimrp commented on src/backend/access/transam/distributedlog.c:
```diff
@@ -69,10 +93,207 @@ static void DistributedLog_SetCommitted(TransactionId localXid,
+	if (oldOldestXmin % (ENTRIES_PER_PAGE * SLRU_PAGES_PER_SEGMENT) !=
+		oldestXmin % (ENTRIES_PER_PAGE * SLRU_PAGES_PER_SEGMENT))
+	{
+		DistributedLog_Truncate();
```
GPDB is notorious in accumulating distributed logs in large volumes.  Truncating distributed log so aggressively is, therefore, quite cool!

It's ok to truncate distributed log without freezing the tuples because:
1. Non-existence of distributed log implies visibility is solely determined by local XID.
AND
2. The QE is guaranteed to receive no distributed snapshot whose in-prgress (or in-future) distributed XIDs fall into the range we just truncated.

If that makes sense, adding a comment as such will help.

asimrp replied:
`xminAllDistributedSnapshots` is monotonically increasing, it can never decrease.  That proves point 2 above.  Does that help ruling out Ashwin's concern?

Eulerizeit commented:
I don't think we should hold this based on performance concerns. Aside from that, if we are sure we won't end in a weird state (Ashwin's question) then we should push.

hlinnaka commented on src/backend/access/transam/distributedlog.c:
```diff
@@ -69,10 +93,207 @@ static void DistributedLog_SetCommitted(TransactionId localXid,
+		{
+			slotno = SimpleLruReadPage(DistributedLogCtl, page, true, xid);
+			entries = (DistributedLogEntry *) DistributedLogCtl->shared->page_buffer[slotno];
+		}
```
Oops, fixed.

hlinnaka commented:
On 07/03/18 12:33, Asim RP wrote:
> The SLRU cache is meant to help backends querying commit status of
> distributed transactions in distributed log.  Using SLRU for
> sequential scans like this affects the LRU bookkeeping adversely, by
> possibly evicting "real recently used" pages unnecessarily.

Well, the LRU bookkeeping will keep the recently used pages in memory, 
regardless of what they're used for. Checking the page that corresponds 
to oldestxmin is just as valid a use of the SLRU as the other queries. I 
don't see a problem here.

(Without a concrete test case that shows a problem, or a concrete 
benchmark that shows that there is no problem, this is all just 
hand-waving.)

- Heikki

hlinnaka commented:
On 06/03/18 20:59, Ashwin Agrawal wrote:
> Just thinking is there possibility to land in weird conditions for
> incorrectly missing distributed log files, as now its not treated as
> ERROR condition but rather driving decisions ?

Yeah, that's a possibility. Or rather, it's a "can't happen" scenario 
that will undoubtedly happen at some point :-).

In this PR, I have a PANIC, if it appears that the oldestXmin in shared 
memory goes past a locally computed oldestXmin. I'm thinking that before 
pushing this, we should downgrade that to an ERROR. In general, any 
confusion or errors around this code should be ephemeral, i.e it should 
fix itself once all the in-progress transactions and snapshots are gone. 
It's good if we can have safeguards to detect apparent confusion, and 
throw ERRORs, but the system as whole should recover.

- Heikki

hlinnaka commented:
Squashed all the bug-fixes, and rebased. This is now ready to be committed, as far as I'm concerned, as soon as the PR pipeline is happy. Any last minute concerns?
